### ISSUE TITLE
Need tests for Get-WinEvent

### ISSUE THREAD

**Issue by iSazonov** [PowerShell/PowerShell#2368]
I need help in creating the tests.

I am ready to fix #2327 but blocked by the lack of tests. 
Now tests for Get-WinEvent is dummy. https://github.com/PowerShell/PowerShell/tree/master/test/powershell/Modules/Microsoft.PowerShell.Diagnostics

The problem is that it is needed to check the "UserData" section of events as a separate case among other tests. This section only is in logs which is empty by default. These logs is:
Microsoft-Windows-AppLocker/EXE and DLL
Microsoft-Windows-AppLocker/MSI and Script
Microsoft-Windows-PrintService/Operational
Microsoft-Windows-PrintService/Admin

As an alternative, I assume that we can exploit the method used in the test of upgrading help system.
https://github.com/PowerShell/PowerShell/blob/master/test/powershell/engine/HelpSystem.Tests.ps1
In other words, create a .evtx file or files with the necessary events and use them instead of real system logs.

What are your thoughts?

**Comment by JamesWTruher** [comment]
I think that if you need specific events to be present, you can't assume that any eventlog file on our CI system will have what you need. I think the best approach is to provide a the specific .evtx file that you need (an additional file for use by a test should go into the `assets` directory where the `Pester` script is.

I'll be adding a few more Get-WinEvent tests shortly

**Comment by iSazonov** [comment]
@JamesWTruher Thanks for the quick help!

Thus  .evtx file.
Here I have one doubt. If we use predetermined file it will be too difficult to maintain and easy to break. The best way to create the evtx on-the-fly from the auxiliary text file containing events in XML format. This requires writing provider on C#. I'm not sure that the solution already exists.

I found another way to meet the task of testing event filters:
>Enable and clear PrintService log
>Create three dummy printers
>Print to the printers
>Make filter test

I suppose it is easier than evtx (but is not so universally).

Samle code:
```
Context 'Get-WinEvent FilterHashtable parameter' { 
        # for this set of tests we need to have a log which has multiple events with UsedData section (ex. PrintService log)

BeforeAll {
   if ( ! $IsWindows ) { return } # @JamesWTruher you dont mentioned in your test guide fot "Context"

   // PrintService/Operational is Disabled by default
   // Required admins rights
   // wevtutil sl <logname> /e:true
   $EventLog = Get-WinEvent -LogName Microsoft-Windows-PrintService/Operational
   if ( $EventLog.IsEnabled )
   {
      $IsEventLogEnabled = $true
   } else
   {
      $EventLog.IsEnabled = $true
      $EventLog.saveChanges()
      $IsEventLog = $false
   }
   # Is this possible in a test environment?
   Clear-EventLog -LogName Microsoft-Windows-PrintService/Operational

   # Create dummy printers
   rundll32 printui.dll,PrintUIEntry /if /b "Test Null Printer 1" /f "%systemroot%\inf\ntprint.inf" /r "nul:" /m "Generic / Text Only" /z
   rundll32 printui.dll,PrintUIEntry /if /b "Test Null Printer 2" /f "%systemroot%\inf\ntprint.inf" /r "nul:" /m "Generic / Text Only" /z
   rundll32 printui.dll,PrintUIEntry /if /b "Test Null Printer 3" /f "%systemroot%\inf\ntprint.inf" /r "nul:" /m "Generic / Text Only" /z

   # Create print events
   "Test text 1" | Out-Printer -Name "Test Null Printer 1"
   "Test text 2" | Out-Printer -Name "Test Null Printer 1"
   "Test text 3" | Out-Printer -Name "Test Null Printer 1"
   "Test text 1" | Out-Printer -Name "Test Null Printer 2"
   "Test text 2" | Out-Printer -Name "Test Null Printer 2"
   "Test text 3" | Out-Printer -Name "Test Null Printer 2"
   "Test text 1" | Out-Printer -Name "Test Null Printer 3"
   "Test text 2" | Out-Printer -Name "Test Null Printer 3"
   "Test text 3" | Out-Printer -Name "Test Null Printer 3"
}

AfterAll {
   # By default Disable PrintService log
   if ( !$IsEventLogEnabled )
   {
      $EventLog.IsEnabled = $false
      $EventLog.saveChanges()
   }

   # Delete dummy printers
   rundll32 printui.dll,PrintUIEntry /dl /n "Test Null Printer 1"
   rundll32 printui.dll,PrintUIEntry /dl /n "Test Null Printer 2"
   rundll32 printui.dll,PrintUIEntry /dl /n "Test Null Printer 3"

}
```

**Comment by lzybkr** [comment]
If a evtx file works on all platforms we support, I think it's reasonable to use a fixed file, but I do like the idea of generating one as well as long as it's easy to control exactly what goes in it.

Your sample uses printers - I don't think that would work everywhere we might test PowerShell, e.g. on Nano.

I have a [simple example](https://gist.github.com/lzybkr/6e4436d300a9050ef398a2f15554fc12) of creating an EventSource in PowerShell. I think it's possible to adapt this sample so it produces events for the event log, but I haven't tried, so that's on approach I'd look into.

The `Windows PowerShell` events are also a good source of events to generate the evtx from, but that could be a tiny bit more fragile depending on how you test and on any changes we might make in generating events.

**Comment by iSazonov** [comment]
@lzybkr Thanks for comments.

Clear about Nano and printers.

Now @JamesWTruher made evtx. This is some trick. And it can only be supported by Powershell team. :confused:

I'd like to see something that the community can support themselves without the ability to break current tests and waiting help from Powershell team.
The simplest way to log needed events is New-EventLog. In this case, we only need to register custom provider which will support events of the desired format (event templates).

**Comment by daxian-dbw** [comment]
Resolved via PR #2373
